firedrake/projection.py

```diff
@@ -129,32 +172,63 @@ def __init__(self, v, v_out, bcs=None, solver_parameters=None, constant_jacobian
         self.v = v
         self.v_out = v_out
         self.bcs = bcs
+        self.method = method
 
-        if not self._same_fspace or self.bcs:
-            V = v_out.function_space()
+        if self.method == "l2":
+            if not self._same_fspace or self.bcs:
+                V = v_out.function_space()
 
-            p = ufl_expr.TestFunction(V)
-            q = ufl_expr.TrialFunction(V)
+                p = ufl_expr.TestFunction(V)
+                q = ufl_expr.TrialFunction(V)
 
-            a = ufl.inner(p, q)*ufl.dx
-            L = ufl.inner(p, v)*ufl.dx
+                a = ufl.inner(p, q)*ufl.dx
+                L = ufl.inner(p, v)*ufl.dx
 
-            problem = vs.LinearVariationalProblem(a, L, v_out, bcs=self.bcs,
-                                                  constant_jacobian=constant_jacobian)
+                problem = vs.LinearVariationalProblem(a, L, v_out, bcs=self.bcs,
+                                                      constant_jacobian=constant_jacobian)
 
-            if solver_parameters is None:
-                solver_parameters = {}
+                if solver_parameters is None:
+                    solver_parameters = {}
 
-            solver_parameters.setdefault("ksp_type", "cg")
+                solver_parameters.setdefault("ksp_type", "cg")
 
-            self.solver = vs.LinearVariationalSolver(problem,
-                                                     solver_parameters=solver_parameters)
+                self.solver = vs.LinearVariationalSolver(problem,
+                                                         solver_parameters=solver_parameters)
+        elif self.method == "average":
+            # NOTE: Any bcs on the function self.v should just work.
+            # Loop over node extent and dof extent
+            V = self.v_out.function_space()
+            shapes = (V.finat_element.space_dimension(), np.prod(V.shape))
+            self._accumulate_kernel = """
+            for (int i=0; i<%d; ++i) {
+                for (int j=0; j<%d; ++j) {
+                    vo[i][j] += v[i][j];
+                    w[i][j] += 1.0;
+            }}""" % shapes
+
+            self._w = function.Function(V)
+        else:
+            raise ValueError("Method type %s not recognized" % str(method))
 
     def project(self):
         """
         Apply the projection.
         """
-        if self._same_fspace and not self.bcs:
-            self.v_out.assign(self.v)
+        if self.method == "l2":
+            if self._same_fspace and not self.bcs:
+                self.v_out.assign(self.v)
+            else:
+                self.solver.solve()
+
         else:
-            self.solver.solve()
+            assert self.method == "average", (
+                "Only 'l2' and 'average' are supported methods at this time."
+            )
```

Comment:
This assertion is always true because you check it in the constructor.  So no need to do so here.